include/triton/Dialect/TritonGPU/Transforms/PipeliningUtility.h

```diff
@@ -50,11 +40,58 @@ void visitNestedOperands(Operation *op, function_ref<void(Value)> visitor);
 /// of `op`.
 SetVector<Value> getNestedOperands(Operation *op);
 
-// Return the minClusterId and maxClusterId for the given ForOp.
-std::pair<int, int> getMinMaxCluster(scf::ForOp &forOp);
-std::pair<int, int> getStageCluster(Operation *op);
-std::optional<std::pair<int, int>> maybeGetStageCluster(Operation *op);
-void setStageCluster(Operation *op, int stage, int cluster);
+// Return maxumum length of the vectorized copy between registers and shared
+// memory for the given tensor type and shared encoding.
+int getCopyVecBytes(RankedTensorType registerTy,
+                    gpu::SharedEncodingTrait sharedEnc);
+
+// Serialize the latencies of the operations in the loops into the latency
+// attribute.
+void serializeLatencies(ModuleOp module, DenseMap<Operation *, int> &opLatency);
+
+// Deserialize the latencies of the operations in the loops from the attribute.
+DenseMap<Operation *, int> deserializeLatencies(ModuleOp module);
+
+// Given a result of MemDescSubview, or Alloca, create a MemDescSubview with a
+// single buffer slice (leading dimension equal to 1), at the given index.
+template <typename BuilderT>
```

Comment:
nit: use OpBuilder base class instead and then move the implementation in the cpp

Reply:
It needs to be a template to use `OpBuilderForStage` which overrides create to set the stage and cluster automatically when you call create.